reactor-netty-http/src/main/java/reactor/netty/http/server/logging/AbstractAccessLogArgProvider.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright (c) 2011-Present VMware, Inc. or its affiliates, All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package reactor.netty.http.server.logging;
+
+import reactor.util.annotation.Nullable;
+
+import java.net.SocketAddress;
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.Objects;
+import java.util.function.Supplier;
+
+/**
+ * @author limaoning
+ */
+abstract class AbstractAccessLogArgProvider<SELF extends AbstractAccessLogArgProvider<SELF>>
+		implements AccessLogArgProvider, Supplier<SELF> {
+
+	static final DateTimeFormatter DATE_TIME_FORMATTER =
+			DateTimeFormatter.ofPattern("dd/MMM/yyyy:HH:mm:ss Z");
+	static final String MISSING = "-";
+
+	final SocketAddress remoteAddress;
+	final String user = MISSING;
+	String zonedDateTime;
+	CharSequence method;
+	CharSequence uri;
+	String protocol;
+	CharSequence status;
+	boolean chunked;
+	long contentLength;
```

Comment:
There are use cases when the I/O handler may send `Transfer-Encoding: chunked` to a HEAD request. As the HEAD request contains only headers, the access log for such request will be with content length `0` which is not quite correct.

`reactor.netty.http.server.AccessLog - /0:0:0:0:0:0:0:1:49307 - - [03/Nov/2020:19:58:38 +0200] "GET / HTTP/1.1" 200 0 10 ms`

This is also valid for HTTP/2
`INBOUND HEADERS: streamId=1 headers=DefaultHttp2Headers[:method: HEAD, :path: /, :scheme: https, :authority: localhost:8080, user-agent: XXX, accept: */*] padding=0 endStream=true`
`OUTBOUND HEADERS: streamId=1 headers=DefaultHttp2Headers[:status: 200] padding=0 endStream=true`

May be we can initialise/clean it `contentLength=-1` so that later it is replaced with `MISSING`?
But please keep in mind that this is relevant only when there is no body, when there is we need to guarantee that we start counting from `0`.

Reply:
Maybe I found a bug, when I defined GET /hello, but client call HEAD /hello, it returns 404, it should returns the GET /hello headers.

Reply:
Do you use `route` if yes then the default implementation does not support declaration of GET but call with HEAD, you need a declaration with HEAD.
For testing purposes you can use `handle`, or custom predicates with `route`.
Check these tests
https://github.com/reactor/reactor-netty/blob/fdd7b1b967fe84f0fa39ae3e57ed0eb41db7b16f/reactor-netty-http/src/test/java/reactor/netty/http/server/HttpServerTests.java#L570
https://github.com/reactor/reactor-netty/blob/fdd7b1b967fe84f0fa39ae3e57ed0eb41db7b16f/reactor-netty-http/src/test/java/reactor/netty/http/server/HttpServerTests.java#L666

Reply:
Should I just ignore the content length like this
`
SELF increaseContentLength(long contentLength) {
	if (chunked && contentLength >= 0 && !HttpMethod.HEAD.asciiName().contentEqualsIgnoreCase(method)) {
		if (this.contentLength == -1) {
			this.contentLength = 0;
		}
		this.contentLength += contentLength;
	}
	return get();
}
`

Reply:
Should I just ignore the content length like this
```
	SELF increaseContentLength(long contentLength) {
		if (chunked && contentLength >= 0 && !HttpMethod.HEAD.asciiName().contentEqualsIgnoreCase(method)) {
			if (this.contentLength == -1) {
				this.contentLength = 0;
			}
			this.contentLength += contentLength;
		}
		return get();
	}
```

Reply:
yes I think this will work

Reply:
```
.route(r -> r.head("/hello",
		(req, res) -> res.header(CONTENT_TYPE, TEXT_PLAIN)
				.sendString(Mono.just("Hello World!"))));
```
But HTTP/2 still returns the content.

Reply:
I'll create an issue for this and we will track it separately. I can reproduce it.